### PR TITLE
all: Transition to using fs watcher by default (fixes #4552)

### LINKED ISSUE #4552: UI for filesystem watcher
Now that we have the watcher code in place, let’s add the UI. There’s a couple of different parts to that:

- enabling it to begin with
- showing its current status
- showing a reasonable error when it doesn’t work

The enabling part is fairly simple. I think we’ve already mostly agreed on a checkbox, and when ticked it will also automatically increase the rescan interval to something high. This happens in the config UI, so the user can change the interval if they want, and in the code it’s still the same interval mechanism as usual. When the watcher is disabled we suggest the usual default interval of 60 seconds. One thing to note here might be a documentation link or comment about why rescans may still be necessary. As far as I know, permission only changes are not picked up by the watcher? Mtime only changes? May need some investigation.

The folder status thing on the left should have a row for the fs watcher status. For most things we don’t show a row unless the value diverges from the default. Nonetheless I think the fs watcher row should always be present and probably replace the rescan interval row. Having it there with “Watcher: On (rescans at 7200s)” or whatever even in the default case gives peace of mind that it’s actually running.  

When it’s enabled but not running we should show an error indication similar to when a folder stopped. (XXX: and change the scan interval...? Not a UI question)

If there are cases when it’s “half running” we need to figure that out. For example if it’s running fine and I copy in a directory into the folder and now it exceeds the system max number of open watch frobbits. 

We should not, at this time, automatically enable watches for existing folders. We may want to have it on by default for newly created folders?

### PR DESCRIPTION
**Pull request by imsodin**
### Purpose

Add the watcher functionality to the edit folder modal, make it the default for new folders and provide a notification to enable it for existing folders. As suggested by @AudriusButkevicius I structured the foler modal with tabs like the settings.

The new default for new folders is watcher enabled and hourly full scans. For existing folders, if the user chooses so in the web UI, watcher is enable and full scans done at 60x current interval capped at once per day.

### Testing

Manual poking around.

### Screenshots

screenshot_2018-03-07_11-23-33

screenshot_2018-03-07_11-23-20

screenshot_2018-03-07_11-23-09

screenshot_2018-03-07_11-21-49

### Documentation

https://github.com/syncthing/docs/pull/379

### CODE REVIEW

**Comment by uok** [comment]
Tabs are a great idea! I think it would also help users make use of the restore feature to enable versioning by default

**Comment by imsodin** [comment]
The issues mentioned in the original PR text were cause by a dirty testing environment and lazy copy&pasting - it's now working as expected.

> I think it would also help users make use of the restore feature to enable versioning by default

That's unrelated to this PR - if you want to discuss it, the forum is the right place.

File: gui/default/syncthing/core/syncthingController.js [hunk 1]
```diff
@@ -62,8 +62,9 @@ angular.module('syncthing.core')
         $scope.folderDefaults = {
             selectedDevices: {},
             type: "readwrite",
-            rescanIntervalS: 60,
+            rescanIntervalS: 3600,
             fsWatcherDelayS: 10,
+            fsWatcherEnabled: true,
```
Comment (AudriusButkevicius) [inline gui/default/syncthing/core/syncthingController.js]: You need an equivalent change in go to have this enabled for introduced folders etc which does not involve js

File: gui/default/syncthing/core/notifications.html [hunk 2]
```diff
@@ -41,3 +41,31 @@ <h3 class="panel-title"><span class="fa fa-flash"></span>&nbsp;<span translate>A
...
+      <p><a href="https://docs.syncthing.net/users/syncing.html#scanning"><span class="fa fa-fw fa-book"></span>&nbsp;<span translate>Learn more</span></a></p>
+      <p>
+        <span translate>Do you want to enable watching for changes for all your folders?</span><br/>
+        <span translate>Additionally the full rescan interval will be increased (times 60, i.e. new default of 1h). You can also configure it manually for every folder later after choosing No.</span>
+      </p>
+      <p translate translate-value-syncthing-inotify="syncthing-inotify">Warning: If you are using an external watcher like {%syncthingInotify%}, you should make sure it is deactivated.</p>
+    </div>
+    <div class="panel-footer clearfix">
+      <div class="pull-right">
+        <button type="button" class="btn btn-primary btn-sm" ng-click="activateAllFsWatchers(); dismiss()">
```
Comment (AudriusButkevicius) [inline gui/default/syncthing/core/notifications.html]: I suspect this might not work due to angular scoping rules.

**Comment by AudriusButkevicius** [comment]
We could also fold ignores modal into this tabbed one.

File: gui/default/syncthing/core/notifications.html (hunk 2 above)
Comment (imsodin) [reply gui/default/syncthing/core/notifications.html]: What do you suspect not to work? It does what I expect it to.

Comment (AudriusButkevicius) [reply gui/default/syncthing/core/notifications.html]: You did test it, and it did work, right?

Comment (imsodin) [reply gui/default/syncthing/core/notifications.html]: Correct, as in the notification vanished and fs watcher was enabled and rescan interval adjusted.

Comment (AudriusButkevicius) [reply gui/default/syncthing/core/notifications.html]: Right, ok.

File: gui/default/syncthing/core/syncthingController.js (hunk 1 above)
Comment (imsodin) [reply gui/default/syncthing/core/syncthingController.js]: done

**Comment by imsodin** [comment]
> We could also fold ignores modal into this tabbed one.

Agreed and already done, but I would like to have this in a separate commit, as its quite a bit of (unrelated) additional diff.

**Comment by uok** [comment]
Are there any downsides to activating the folder watcher? If yes, this should also be mentioned in the yes/now notification.

**Comment by imsodin** [comment]
@canton7 You already have some transition mechanism in place, right? Therefore you probably don't want the notification about enabling the watcher displayed - so you'll need to remove "fsWatcherNotification" from `unackedNotificationID` on upgrade, is that ok?

@kozec @sieren @Martchus Pinging you in case your wrappers do anything about filesystem notifications.

> Are there any downsides to activating the folder watcher? If yes, this should also be mentioned in the yes/now notification.

None are known to me (there likely are scenarios where a watcher isn't wanted, but that's not really a downside). I assumed by stating that the watcher is the new default, it is sufficiently clear that it is recommended for use, isn't it?

**Comment by canton7** [comment]
> @canton7 You already have some transition mechanism in place, right? Therefore you probably don't want the notification about enabling the watcher displayed - so you'll need to remove "fsWatcherNotification" from unackedNotificationID on upgrade, is that ok?

My strategy is simple: if the user enables filesystem watching in Syncthing, I disable mine. So I'm happy to leave your work here untouched: as users switch to using your filesystem watcher, mine will get out of the way.

In future I'll need some strategy to move users over to yours as I deprecate mine, but that's in the dim and distant future.

**Comment by imsodin** [comment]
By accident I discovered a very strange race: Since enabling the watcher by default in go I get the following weird race in model *only if fswatcher debug facility is activated*:
```
WARNING: DATA RACE
Read at 0x00c42caac7f0 by goroutine 331:
  reflect.typedmemmove()
      /media/ext4_data/Linux/source/go/src/runtime/mbarrier.go:259 +0x0
  reflect.packEface()
      /media/ext4_data/Linux/source/go/src/reflect/value.go:112 +0x10b
  reflect.valueInterface()
      /media/ext4_data/Linux/source/go/src/reflect/value.go:961 +0x170
  reflect.Value.Interface()
      /media/ext4_data/Linux/source/go/src/reflect/value.go:931 +0x51
  fmt.(*pp).printValue()
      /media/ext4_data/Linux/source/go/src/fmt/print.go:692 +0x3994
  fmt.(*pp).printValue()
      /media/ext4_data/Linux/source/go/src/fmt/print.go:846 +0x26da
  fmt.(*pp).printArg()
      /media/ext4_data/Linux/source/go/src/fmt/print.go:682 +0x19e
  fmt.(*pp).doPrintln()
      /media/ext4_data/Linux/source/go/src/fmt/print.go:1136 +0x6e
  fmt.Sprintln()
      /media/ext4_data/Linux/source/go/src/fmt/print.go:264 +0x5f
  github.com/syncthing/syncthing/lib/logger.(*logger).debugln()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/logger/logger.go:114 +0xa8
  github.com/syncthing/syncthing/lib/logger.(*facilityLogger).Debugln()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/logger/logger.go:284 +0xc6
  github.com/syncthing/syncthing/lib/fs.(*logFilesystem).Watch()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/fs/logfs.go:133 +0x515
  github.com/syncthing/syncthing/lib/model.(*folder).startWatchAsync()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:212 +0x291

Previous write at 0x00c42caac7f0 by goroutine 392:
  github.com/syncthing/syncthing/lib/ignore.(*Matcher).parseLocked()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/ignore/ignore.go:159 +0x1e7
  github.com/syncthing/syncthing/lib/ignore.(*Matcher).Load()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/ignore/ignore.go:137 +0x2b5
  github.com/syncthing/syncthing/lib/model.(*Model).internalScanFolderSubdirs()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:1942 +0x5d5
  github.com/syncthing/syncthing/lib/model.(*folder).scanSubdirs()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:131 +0xcb
  github.com/syncthing/syncthing/lib/model.(*folder).scanTimerFired()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:143 +0x6e
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/rwfolder.go:201 +0xae6
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:525 +0x70

Goroutine 331 (running) created at:
  github.com/syncthing/syncthing/lib/model.(*folder).startWatch()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:202 +0x2c1
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/rwfolder.go:156 +0xebb
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:525 +0x70

Goroutine 392 (running) created at:
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).runService()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:515 +0x68
  github.com/syncthing/syncthing/vendor/github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/vendor/github.com/thejerf/suture/supervisor.go:413 +0xda6
==================
```
This seems completely bogus - any ideas as to why and/or how to get rid of it?

**Comment by calmh** [comment]
Not bogus, but not super serious. The logfs is doing bad stuff here as it’s passing the *ignore.Matcher directly to Debugln(), which traverses the whole ignore matcher struct using reflect without holding any locks, which is racy. That log line in logfs.go#133 should be fixed.

File: gui/default/assets/lang/lang-en.json [hunk 3]
```diff
@@ -49,6 +50,7 @@
    "Connection Error": "Connection Error",
    "Connection Type": "Connection Type",
    "Connections": "Connections",
+   "Contiuously watching for changes is now available within Syncthing. This will detect changes on disk and issue a scan on only the modified paths. The benefits are that changes are propagated quicker and that less full scans are required.": "Contiuously watching for changes is now available within Syncthing. This will detect changes on disk and issue a scan on only the modified paths. The benefits are that changes are propagated quicker and that less full scans are required.",
```
Comment (calmh) [inline gui/default/assets/lang/lang-en.json]: Spelling, Continuously. And I think s/less/fewer/

**Comment by calmh** [comment]
Any chance gets to the rescan interval in th folder editor dialog when checking/unchecking the watcher? I think we discussed setting it to 60/3600 in response to the checkbox, and letting the user edit thereafter?

**Comment by imsodin** [comment]
@calmh The interval is now set to 60/3600 on toggling the watcher checkbox.  
@AudriusButkevicius Ping regarding your review.

I think it would make sense to have this in the next release, now that the watcher status is visible in the "main folder panes". Any showstoppers on this?

**Comment by calmh** [comment]
I'm actually going to go negatory on that and say that .46 will do well as just a bugfix release. We can introduce new regressions in .47 instead, including the fallout from turning on fswatcher by default.

**Comment by AudriusButkevicius** [comment]
I'm happy with this.

It seems @calmh is too, if not, we can always revert.

**Pull request merged**